feat: Add functionality to mark conversations as read in Chatwoot

bergpinheiro opened the pull request:
## 📋 Description

This PR adds the ability to mark conversations as "read" in the WAHA-Chatwoot integration.

## ✨ Features

- ✅ New `WAHAMessageReadConsumer` consumer to process `MESSAGE_ACK` events
- ✅ Filter to process only read confirmations (`ack === 3`)
- ✅ Integration with Chatwoot API using `update_last_seen` endpoint
- ✅ Webhook filter to avoid infinite loops (ignores `status: "read"`)
- ✅ Full support for contact and conversation mapping

## 🔧 Technical Changes

### New Files:
- `src/apps/chatwoot/consumers/waha/message.ack.ts` - Consumer for read events

### Modified Files:
- `src/apps/chatwoot/chatwoot.module.ts` - Registration of the new queue and consumer
- `src/apps/chatwoot/client/ConversationService.ts` - `markAsRead()` method
- `src/apps/chatwoot/client/ContactConversationService.ts` - `markConversationAsRead()` method
- `src/apps/chatwoot/consumers/waha/base.ts` - Added `MESSAGE_ACK` to events
- `src/apps/chatwoot/services/ChatWootWAHAQueueService.ts` - Routing to new queue
- `src/apps/chatwoot/api/chatwoot.webhook.controller.ts` - Webhook filter
- `src/structures/enums.dto.ts` - Added `MESSAGE_READ` enum

## 🧪 How to Test

1. Send a message from Chatwoot to WhatsApp
2. Read the message in WhatsApp
3. Verify that the The conversation was marked as read in Chatwoot.

## 📝 Notes

- Uses `MESSAGE_ACK` instead of `MESSAGE_READ` (which is not generated by WAHA)
- Correct URL: `/public/api/v1/inboxes/{inbox_identifier}/contacts/{source_id}/conversations/{conversation_id}/update_last_seen`
- Authentication using `inboxIdentifier` as a token
- Webhook filter to avoid processing read receipts

## 🔗 Related

- Resolves the need to mark conversations as read in the Chatwoot integration
- Improves the user experience with real-time read status

Related to https://github.com/devlikeapro/waha/issues/1319

[![patron:PRO](https://img.shields.io/badge/patron-PRO-188a42)](https://waha.devlike.pro/docs/how-to/plus-version/#tiers)

bergpinheiro committed:
feat(chatwoot): implement message read status functionality

- Add WAHAMessageReadConsumer to process MESSAGE_ACK events
- Filter to process only read confirmations (ack === 3)
- Integration with Chatwoot API using update_last_seen endpoint
- Webhook filter to avoid infinite loops (ignores status: 'read')
- Full support for contact and conversation mapping

New Files:
- src/apps/chatwoot/consumers/waha/message.read.ts - Consumer for read events

Modified Files:
- src/apps/chatwoot/chatwoot.module.ts - Registration of new queue and consumer
- src/apps/chatwoot/client/ConversationService.ts - markAsRead() method
- src/apps/chatwoot/client/ContactConversationService.ts - markConversationAsRead() and getSourceIdByChatId() methods
- src/apps/chatwoot/consumers/waha/base.ts - Added MESSAGE_ACK to events
- src/apps/chatwoot/services/ChatWootWAHAQueueService.ts - Routing to new queue
- src/apps/chatwoot/api/chatwoot.webhook.controller.ts - Webhook filter
- src/apps/chatwoot/consumers/QueueName.ts - Added WAHA_MESSAGE_READ enum

Features:
âœ… Real-time read status between WhatsApp and Chatwoot
âœ… Improved user experience with read confirmations
âœ… Robust integration using official Chatwoot APIs
âœ… Infinite loop prevention with smart filters
âœ… Detailed logging for debugging and monitoring

Resolves the need to mark conversations as read in the Chatwoot integration
Improves the user experience with real-time read status

devlikepro commented on src/apps/chatwoot/consumers/waha/message.ack.ts:
We use event in consumer name - so it'd be `message.ack.ts` because it process `message.ack` event.

devlikepro reviewed (approved):
requires a bit of changes, but overall it's gooooood :tada:

bergpinheiro committed:
refactor: rename message.read.ts to message.ack.ts for better naming convention

- Rename file to reflect the MESSAGE_ACK event it processes
- Update import reference in chatwoot.module.ts
- Maintains same functionality while improving code clarity

bergpinheiro committed:
Merge remote-tracking branch 'upstream/core' into feature-only-last

gabrielnogueiracsc commented:
+1

devlikepro commented on src/apps/chatwoot/api/chatwoot.webhook.controller.ts:
```diff
@@ -46,6 +46,11 @@ export class ChatwootWebhookController {
     }
 
+    // Ignore messages with status "read" (from update_last_seen endpoint)
+    if (body.status === 'read') {
```
There's no such field :( 
The only `message_updated` we need - when agent clicks on "Retry" button in CHatWoot message.
Found a bit weird way to identify it, hopefully it'll work :crossed_fingers:
